### PR TITLE
detect: allow bypass keyword for fw rules in firewall mode (only) - v3

### PR DESCRIPTION
**Pull request by jufajardini**
Link to ticket: https://redmine.openinfosecfoundation.org/issues/
https://redmine.openinfosecfoundation.org/issues/8459

Previous PR: https://github.com/OISF/suricata/pull/15955

Describe changes:
- document bypass keyword for firewall mode, including impact on stats, accepted hooks, and expectations with `pre_flow` hook.
- better commit segmentation (hopefully)
- `accept+bypass` packet isn't counted as `ips.accepted` in stats (added a new packet flag to indicate firewall's bypass, for this

TODOs?
- tests and revAIwing iterations uncovered that:
 -  thresholding could suppress firewall alerts -- independent of this work, but must decide what to do
 - pre-stream hook is still a corner case
 - pre-flow rules are apparently added to two different SGH, leading to two alerts, when the action isn't drop/ config
(the last two are visible through sv tests *-152 and *-153)

SV_BRANCH=https://github.com/OISF/suricata-verify/pull/3268

### CODE REVIEW

File: rust/sys/src/sys.rs [hunk 1]
```diff
@@ -24,6 +24,13 @@ pub const SIGMATCH_INFO_ENUM_UINT: u32 = 524288;
 pub const SIGMATCH_INFO_BITFLAGS_UINT: u32 = 1048576;
 pub const SIGMATCH_BAN_FIREWALL_RULE: u32 = 2097152;
 pub const SIGMATCH_BAN_FIREWALL_MODE: u32 = 4194304;
+pub const SIGMATCH_BAN_TD_FIREWALL_MODE: u32 = 8388608;
+pub const SIGMATCH_BAN_FIREWALL_SCOPE_PACKET: u32 = 16777216;
+pub const SIGMATCH_BAN_FIREWALL_SCOPE_TX: u32 = 33554432;
+pub const SIGMATCH_BAN_FIREWALL_SCOPE_HOOK: u32 = 67108864;
+pub const SIGMATCH_BAN_ACTION_CONFIG: u32 = 134217728;
+pub const SIGMATCH_BAN_ACTION_DROP: u32 = 268435456;
+pub const SIGMATCH_BAN_ACTION_REJECT: u32 = 536870912;
```
Comment (jufajardini) [inline rust/sys/src/sys.rs:33]: I think these have to also include `FIREWALL`, too.

File: src/detect-engine-register.h [hunk 2]
```diff
@@ -354,6 +354,16 @@ extern int DETECT_TBLSIZE_IDX;
...
+/** keyword cannot be used in td rules with firewall mode */
+#define SIGMATCH_BAN_TD_FIREWALL_MODE (1UL << (23))
+/** keyword cannot be used in combination with indicated action scope */
+#define SIGMATCH_BAN_FIREWALL_SCOPE_PACKET (1UL << (24))
+#define SIGMATCH_BAN_FIREWALL_SCOPE_TX     (1UL << (25))
+#define SIGMATCH_BAN_FIREWALL_SCOPE_HOOK   (1UL << (26))
+/** keyword cannot be unsed in combination with indicated action */
+#define SIGMATCH_BAN_ACTION_CONFIG (1UL << (27))
+#define SIGMATCH_BAN_ACTION_DROP   (1UL << (28))
+#define SIGMATCH_BAN_ACTION_REJECT (1UL << (29))
```
Comment (jufajardini) [inline src/detect-engine-register.h:366]: Same as with the rust bindings, as these are the original ones.

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline = 32876

File: doc/userguide/rules/bypass-keyword.rst [hunk 3]
```diff
@@ -26,3 +30,49 @@ Bypass a flow on matching http traffic.
...
+.. note:: If `bypass` is used in a rule together with thresholding, the bypass
+   could be silent, if the alert is suppressed.
+
+Valid firewall rule with bypass:
+
+.. container:: example-rule
+
+  :example-rule-emphasis:`accept:flow,alert` http1:request_headers any any -> \
+  any any (http.host; content:"suricata.io"; :example-rule-emphasis:`bypass;` \
+  sid:10001; rev:1;)
```
Comment (jufajardini) [inline doc/userguide/rules/bypass-keyword.rst:64]: CI docs build failing for not recognizing "flow,alert" as a valid action scope.

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline = 32886

File: doc/userguide/rules/bypass-keyword.rst (hunk 3 above)
Comment (jufajardini) [reply doc/userguide/rules/bypass-keyword.rst:64]: https://github.com/OISF/suricata/pull/16052 Should fix this failure.

File: src/detect-engine-register.h (hunk 2 above)
Comment (jufajardini) [reply src/detect-engine-register.h:366]: Missing: ACTION_PASS. Should it be accepted, or banned?

File: doc/userguide/rules/bypass-keyword.rst [hunk 4]
```diff
@@ -26,3 +30,49 @@ Bypass a flow on matching http traffic.
...
+
+.. attention:: `bypass` on a firewall rule is terminal. Threat detection rules
+  are not evaluated for the matching packet, respecting the premise of what would happen if Firewall and IPS were two separate devices.
+
+.. note:: The type of bypass will depend on whether the engine is configured
+  for local or capture bypass: offloading is not guaranteed by a firewall
+  bypass rule.
+
+.. note:: If `bypass` is used in a rule together with thresholding, the bypass
+   could be silent, if the alert is suppressed.
```
Comment (jufajardini) [inline doc/userguide/rules/bypass-keyword.rst:56]: Do we want more than this? eg banning thresholding from use with bypass? We may need to document stats hole...

**Comment by jufajardini** [comment]
Followed by: https://github.com/OISF/suricata/pull/16063
